blockchain/blockchain.go

```diff
@@ -444,6 +447,17 @@ func (bc *BlockChain) StateAt(root common.Hash) (*state.StateDB, error) {
 	return state.New(root, bc.stateCache)
 }
 
+// StateAtWithGCLock returns a new mutable state based on a particular point in time with read lock of the state nodes.
+func (bc *BlockChain) StateAtWithGCLock(root common.Hash) (*state.StateDB, error) {
+	bc.LockGCCachedNode()
+	exist := bc.stateCache.TrieDB().DoesExistCachedNode(root)
+	if exist {
+		return state.New(root, bc.stateCache)
```

Comment:
What if `state.New` returns an error? We don't unlock here, but the caller also doesn't unlock the lock.

Reply:
Thanks. Reflected. :)